### DISCUSSION TITLE
Use Inject outside class

### DISCUSSION THREAD

**Discussion by lgmf** [TypedProject/ts-express-decorators#2036] [Q&A] [answered]
```js
import {UseAfter} from "@tsed/platform-middlewares";
import {useDecorators} from "@tsed/core";

export function CustomStatus(code: number) {
  return useDecorators(
    UseAfter((req: any, res: any, next: any) => {
      res.status(code);
      next();
    })
  );
}
```

I've seen this example on the docs.

How can I use the DI in this case?

**Comment by Romakita** [comment]
Why?

**Reply by Romakita** [reply]
Use a middleware class id you need to use injector.

this example is just to use an express middleware without a huge logic.

see you

**Reply by lgmf** [reply]
In order to pass custom arguments

I wanna create a middleware to check for roles

@CheckRole("admin')
@CheckRole("editor");

And using the StoreSet 

```js
export function AcceptMime(...mimes: string[]): Function {
  return useDecorators(StoreSet(PlatformAcceptMimesMiddleware, mimes), UseBefore(PlatformAcceptMimesMiddleware));
}
```

it only work if I use the middleware on the endpoint, If I use it global on the controller it doesn't work

**Reply by lgmf** [reply]
I was thinking about something like we have when testing

```
      const service = PlatformTest.get<ParseService>(ParseService);
```

**Reply by Romakita** [reply]
You mix differents topics… 

UseAfter applied on controller or endpoint have always access to the endpoint.

It’s the same thing for UseBeforeEach used on controller. Also the middleware will always applied on endpoints. Have you read this part of the doc? 

https://tsed.io/docs/middlewares.html#call-sequences

It explain the difference between all decorators.

see you

**Reply by lgmf** [reply]
Let me try to explain the problem better then.

Considering that I have this middleware
```js
import { Context, Middleware, MiddlewareMethods, Next, UseBefore } from "@tsed/common";
import { StoreSet, useDecorators } from "@tsed/core";

@Middleware()
class MyMiddlewareImpl implements MiddlewareMethods {
  async use(@Context() ctx: Context, @Next() next: any) {
    const store = ctx.endpoint.store;
    const code = store.get("code");

    console.log("code", code);

    next();
  }
}

export function MyMiddleware(code: number) {
  return useDecorators(StoreSet("code", code), UseBefore(MyMiddlewareImpl));
}
```

If I use it like this

```js
@Controller("/app")
export class MyController {

  @Get()
  @MyMiddleware(201)
  async myEndpoint() {
    return "hello world!";
  }
}
```
It prints "code 201".

but if I use it like this

```js
@Controller("/app")
@MyMiddleware(201)
export class MyController {

  @Get()
  async myEndpoint() {
    return "hello world!";
  }
}
```

It prints "code undefined".

**Reply by Romakita** [reply]
This exactly that I explained in my previous comment…
Use the UseBeforeEach decorator instead UseBefore.

see you

**Reply by lgmf** [reply]
Like this

```
export function MyMiddleware(code: number) {
  return useDecorators(StoreSet("code", code), UseBeforeEach(MyMiddlewareImpl));
}
```

?

**Reply by Romakita** [reply]
Yes

**Reply by lgmf** [reply]
I'm still receiving undefined on the case below:

```
@Controller("/app")
@MyMiddleware(201)
export class MyController {

  @Get()
  async myEndpoint() {
    return "hello world!";
  }
}
```

https://replit.com/@lgmferrari/DarkorchidNoteworthyAmoeba

**Reply by lgmf** [reply]
I wrote this test file on the replit I've sent you on my last comment

```js
import { PlatformTest } from "@tsed/common";
import SuperTest from "supertest";
import { HelloWorldController } from "./HelloWorldController";
import { Server } from "../../Server";

describe("HelloWorldController", () => {
  let request: SuperTest.SuperTest<SuperTest.Test>;

  beforeEach(PlatformTest.bootstrap(Server, {
    mount: {
      "/": [HelloWorldController]
    }
  }));

  beforeEach(() => {
    request = SuperTest(PlatformTest.callback());
  });

  afterEach(PlatformTest.reset);

  it("/test/works: should return the correct code", async () => {
    const { body, status } = await request.get('/test/works');
    expect(status).toBe(200);
    expect(body.code).toBe(777)
  });

  it("/test/doesnt-work: should return the correct code", async () => {
    const { body, status } = await request.get('/test/doesnt-work');
    expect(status).toBe(200);
    expect(body.code).toBe(999);
  });
});
```

![image](https://user-images.githubusercontent.com/8532045/184353654-01be8494-20a0-42f9-9d15-ad0383cf087c.png)


That's why I was try to find another way to pass custom arguments to a middleware and that's why I asked you how to use tsed DI outside the classes

**Reply by Romakita** [reply]
Hello @lgmf 

Sorry for my wrong answer. It wasn't easy to give you the right response on my mobile phone. And thanks for your patience and thanks also for the repo example!

Here is the correct code:

```typescript
import {Context, Middleware, MiddlewareMethods, UseBeforeEach} from "@tsed/common";
import {StoreSet, useDecorators} from "@tsed/core";

interface MyMiddlewareParams {
  code: number;
}

@Middleware()
class MyMiddlewareImpl implements MiddlewareMethods {
  static params: MyMiddlewareParams;

  async use(@Context() ctx: Context) {
    const code = ctx.endpoint.store.get("code") || ctx.endpoint.parent.store.get("code");

    ctx.set("code", code);
  }
}

export function MyMiddleware(code: number) {
  return useDecorators(StoreSet("code", code), UseBeforeEach(MyMiddlewareImpl));
}
```
The fix is around this code: `const code = ctx.endpoint.store.get("code") || ctx.endpoint.parent.store.get("code");`

`SetStore` is always contextualized. So if you use `SetStore` on top of the class. The metadata is stored on `target`. But if you use it on endpoint, the metadata is automatically attached to the `EndpointMetadata.store`.

See you
Romain

**Reply by Romakita** [reply]
Optimization:

```typescript
@Middleware()
class MyMiddlewareImpl implements MiddlewareMethods {
  static params: MyMiddlewareParams;

  async use(@Context() ctx: Context) {
    const code = ctx.endpoint.store.get("code") || ctx.endpoint.parent.store.get("code");

    ctx.set("code", code);
  }
}
```
Use Next decorator is not necessary in most case because Ts.ED handle async method/Observable/Stream. Next is just here to wrap an Express middleware. But even then, it's not necessary.

**Reply by lgmf** [reply]
No problem at all!

Got it.

Thanks a lot for the help :D 

![image](https://user-images.githubusercontent.com/8532045/184361653-9712a9a8-c1de-444a-a603-5f1a218dbf58.png)

**Reply by Romakita** [reply]
You can mark this topic as answered ;)

**Reply by lgmf** [reply]
Done! 

If you wanna paste the answer in a different comment to make it easier to find I can mark it later as the answer

**Reply by Romakita** [reply]
@lgmf done ;)

**Comment by Romakita** [comment, marked as answer]
# Use a middleware on class and/or endpoint and store data

This example show you, how you can create a custom decorator to collect metadata on class and/or on class method and register a middleware.

```typescript
import {Context, Middleware, MiddlewareMethods, UseBeforeEach} from "@tsed/common";
import {StoreSet, useDecorators} from "@tsed/core";

interface MyMiddlewareParams {
  code: number;
}

@Middleware()
class MyMiddlewareImpl implements MiddlewareMethods {
  static params: MyMiddlewareParams;

  async use(@Context() ctx: Context) {
    const code = ctx.endpoint.store.get("code") || ctx.endpoint.parent.store.get("code");

    ctx.set("code", code);
  }
}

export function MyMiddleware(code: number) {
  return useDecorators(StoreSet("code", code), UseBeforeEach(MyMiddlewareImpl));
}
```
> The trick is around this code: `const code = ctx.endpoint.store.get("code") || ctx.endpoint.parent.store.get("code");`

`SetStore` is always contextualized. So if you use `SetStore` on top of the class. The metadata is stored on target. But if you use it on endpoint, the metadata is automatically attached to the `EndpointMetadata.store`.

## Usage on controller

```typescript
import {Controller} from "@tsed/di";
import {Get} from "@tsed/schema";
import {Context} from "@tsed/common";

import {MyMiddleware} from "../../middlewares/MyMiddleware";

@Controller("/test")
@MyMiddleware(999) // by default
export class HelloWorldController {

  @Get("/scenario-1")
  @MyMiddleware(777) // specific to the current endpoint
  work(@Context("code") code: number) {
    return {code}; // 777
  }

  @Get("/scenario-2")
  doesntWork(@Context("code") code: number) {
    return {code};  // 999 - value set on controller
  }
}
```

See you
Romain

**Reply by lgmf** [reply]
It worked!

Thanks
